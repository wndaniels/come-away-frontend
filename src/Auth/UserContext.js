import React from "react";
/**
 *  Context:
 *
 * Provides currentUser object and setter for it throughout app.
 */
const UserContext = React.createContext();

export default UserContext;
