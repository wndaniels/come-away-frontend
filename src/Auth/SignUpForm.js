import Alert from "../Common/Alert.js";
import { useContext, useState } from "react";
import { useNavigate } from "react-router-dom";
import { Form } from "react-router-dom";
import UserContext from "./UserContext.js";

const SignUpForm = () => {
  const { signup } = useContext(UserContext);
  const navigate = useNavigate();
  const [formData, setFormData] = useState({
    username: "",
    password: "",
    firstName: "",
    lastName: "",
    email: "",
  });
  const [formErr, setFormErr] = useState([]);

  /**
   * Handles Form Submit:
   *
   * Calls signup function and if successful, navigates to homepage.
   */
  async function handleSubmit(evt) {
    evt.preventDefault();
    let result = await signup(formData);
    if (result.success) {
      navigate("/");
    } else {
      setFormErr(result.errors);
    }
  }

  /**
   * Handles update for form data fields.
   */
  const handleChange = (evt) => {
    const { name, value } = evt.target;
    setFormData((l) => ({ ...l, [name]: value }));
  };

  return (
    <div className="SignUpForm m-5">
      <div className="container col-md-6 offset-md-3 col-lg-4 offset-lg-4">
        <h1 className="mb-3">Sign Up</h1>
        <div className="card">
          <div className="card-body">
            <Form method="post" onSubmit={handleSubmit}>
              <div className="d-grid gap-3">
                <div className="form-group">
                  <input
                    type="text"
                    name="username"
                    placeholder="Username"
                    className="form-control"
                    value={formData.username}
                    onChange={handleChange}
                  />
                </div>
                <div className="form-group">
                  <input
                    type="password"
                    name="password"
                    placeholder="Password"
                    className="form-control"
                    value={formData.password}
                    onChange={handleChange}
                  />
                </div>
                <div className="form-group">
                  <input
                    type="text"
                    name="firstName"
                    placeholder="First Name"
                    className="form-control"
                    value={formData.firstName}
                    onChange={handleChange}
                  />
                </div>
                <div className="form-group">
                  <input
                    type="text"
                    name="lastName"
                    placeholder="Last Name"
                    className="form-control"
                    value={formData.lastName}
                    onChange={handleChange}
                  />
                </div>
                <div className="form-group">
                  <input
                    type="email"
                    name="email"
                    placeholder="Email"
                    className="form-control mb-3"
                    value={formData.email}
                    onChange={handleChange}
                  />
                </div>
              </div>

              {formErr.length ? (
                <Alert type="danger" messages={formErr} />
              ) : null}

              <button
                onSubmit={handleSubmit}
                className="btn btn-sm btn-primary"
              >
                Sign Up
              </button>
            </Form>
          </div>
        </div>
      </div>
    </div>
  );
};

export default SignUpForm;
