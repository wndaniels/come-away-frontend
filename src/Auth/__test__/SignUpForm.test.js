import React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import "@testing-library/jest-dom";
import SignUpForm from "../SignUpForm";
import UserContext from "../UserContext";

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => jest.fn(),
}));

describe("SignUpForm", () => {
  it("renders without crashing", () => {
    const { asFragment } = render(
      <Wrapper>
        <SignUpForm />
      </Wrapper>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("can find placeholder", () => {
    const { getByPlaceholderText } = render(
      <Wrapper>
        <SignUpForm />
      </Wrapper>
    );
    const usernamePlaceholder = getByPlaceholderText(/username/i);
    expect(usernamePlaceholder).toBeInTheDocument();
  });
});

function Wrapper({ children }) {
  return (
    <UserContext.Provider
      value={{
        currentUser: jest.fn(),
        setCurrentUser: jest.fn(),
        login: jest.fn(),
        signup: jest.fn(),
        logout: jest.fn(),
      }}
    >
      {children}
    </UserContext.Provider>
  );
}
