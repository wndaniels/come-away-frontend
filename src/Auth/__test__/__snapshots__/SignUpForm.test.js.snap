// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpForm renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="SignUpForm m-5"
  >
    <div
      class="container col-md-6 offset-md-3 col-lg-4 offset-lg-4"
    >
      <h1
        class="mb-3"
      >
        Sign Up
      </h1>
      <div
        class="card"
      >
        <div
          class="card-body"
        >
          <form
            method="post"
          >
            <div
              class="d-grid gap-3"
            >
              <div
                class="form-group"
              >
                <input
                  class="form-control"
                  name="username"
                  placeholder="Username"
                  type="text"
                  value=""
                />
              </div>
              <div
                class="form-group"
              >
                <input
                  class="form-control"
                  name="password"
                  placeholder="Password"
                  type="password"
                  value=""
                />
              </div>
              <div
                class="form-group"
              >
                <input
                  class="form-control"
                  name="firstName"
                  placeholder="First Name"
                  type="text"
                  value=""
                />
              </div>
              <div
                class="form-group"
              >
                <input
                  class="form-control"
                  name="lastName"
                  placeholder="Last Name"
                  type="text"
                  value=""
                />
              </div>
              <div
                class="form-group"
              >
                <input
                  class="form-control mb-3"
                  name="email"
                  placeholder="Email"
                  type="email"
                  value=""
                />
              </div>
            </div>
            <button
              class="btn btn-sm btn-primary"
            >
              Sign Up
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
